File: typescript/tests/guardrails-step-enforcer.test.ts
import { StepEnforcer, StepNotAllowedError } from "../src/middleware/guardrails/step_enforcer";

describe("StepEnforcer", () => {
  describe("addStep", () => {
    it("should add a step with prerequisites", () => {
      const se = new StepEnforcer();
      se.addStep("deploy", ["build", "test"]);

      const [allowed, missing] = se.canExecute("session1", "deploy");
      expect(allowed).toBe(false);
      expect(missing).toHaveLength(2);
    });
  });

  describe("markStepComplete", () => {
    it("should mark a step as complete", () => {
      const se = new StepEnforcer();
      se.addStep("deploy", ["build", "test"]);

      se.markStepComplete("session1", "build");
      const [allowed, missing] = se.canExecute("session1", "deploy");

      expect(allowed).toBe(false);
      expect(missing).toContain("test");
    });
  });

  describe("validateExecution", () => {
    it("should not throw when prerequisites are met", () => {
      const se = new StepEnforcer();
      se.addStep("deploy", ["build"]);

      se.markStepComplete("session1", "build");
      expect(() => se.validateExecution("session1", "deploy")).not.toThrow();
    });

    it("should throw when prerequisites are not met", () => {
      const se = new StepEnforcer();
      se.addStep("deploy", ["build"]);

      expect(() => se.validateExecution("session1", "deploy")).toThrow(StepNotAllowedError);
    });
  });

  describe("resetSession", () => {
    it("should reset completed steps for a session", () => {
      const se = new StepEnforcer();
      se.addStep("deploy", ["build"]);

      se.markStepComplete("session1", "build");
      se.resetSession("session1");

      const [allowed] = se.canExecute("session1", "deploy");
      expect(allowed).toBe(false);
    });
  });

  describe("isTerminalAllowed", () => {
    it("should return true when terminal is allowed", () => {
      const se = new StepEnforcer();
      se.addStep("deploy", ["build"]);
      se.addStep("build", []);

      se.markStepComplete("session1", "build");
      const allowed = se.isTerminalAllowed("session1", "deploy");

      expect(allowed).toBe(true);
    });
  });

  describe("getAllowedTerminals", () => {
    it("should return allowed terminals", () => {
      const se = new StepEnforcer();
      se.addStep("deploy", ["build"]);
      se.addStep("test", []);

      const allowed = se.getAllowedTerminals("session1");

      expect(allowed).toContain("test");
      expect(allowed).not.toContain("deploy");
    });
  });

  describe("no prerequisites", () => {
    it("should allow tools with no prerequisites", () => {
      const se = new StepEnforcer();
      se.addStep("build", []);

      const [allowed] = se.canExecute("session1", "build");
      expect(allowed).toBe(true);
    });
  });

  describe("invalid session", () => {
    it("should not allow for nonexistent session", () => {
      const se = new StepEnforcer();
      se.addStep("deploy", ["build"]);

      const [allowed] = se.canExecute("nonexistent", "deploy");
      expect(allowed).toBe(false);
    });
  });
});

File: python/tests/guardrails_step_enforcer_test.py
import pytest

from pedro_agentware.middleware.guardrails.step_enforcer import (
    StepEnforcer,
    StepNotAllowedError,
)


def test_add_step():
    se = StepEnforcer()
    se.add_step("deploy", ["build", "test"])

    allowed, missing = se.can_execute("session1", "deploy")
    assert allowed is False
    assert len(missing) == 2


def test_mark_step_complete():
    se = StepEnforcer()
    se.add_step("deploy", ["build", "test"])

    se.mark_step_complete("session1", "build")
    allowed, missing = se.can_execute("session1", "deploy")

    assert allowed is False
    assert missing == ["test"]


def test_validate_execution_success():
    se = StepEnforcer()
    se.add_step("deploy", ["build"])

    se.mark_step_complete("session1", "build")
    se.validate_execution("session1", "deploy")


def test_validate_execution_failure():
    se = StepEnforcer()
    se.add_step("deploy", ["build"])

    with pytest.raises(StepNotAllowedError) as exc_info:
        se.validate_execution("session1", "deploy")

    assert exc_info.value.tool == "deploy"
    assert "build" in exc_info.value.missing_steps


def test_reset_session():
    se = StepEnforcer()
    se.add_step("deploy", ["build"])

    se.mark_step_complete("session1", "build")
    se.reset_session("session1")

    allowed, _ = se.can_execute("session1", "deploy")
    assert allowed is False


def test_is_terminal_allowed():
    se = StepEnforcer()
    se.add_step("deploy", ["build"])
    se.add_step("build", [])

    se.mark_step_complete("session1", "build")
    allowed = se.is_terminal_allowed("session1", "deploy")

    assert allowed is True


def test_get_allowed_terminals():
    se = StepEnforcer()
    se.add_step("deploy", ["build"])
    se.add_step("test", [])

    allowed = se.get_allowed_terminals("session1")

    assert allowed == ["test"]


def test_no_prerequisites():
    se = StepEnforcer()
    se.add_step("build", [])

    allowed, _ = se.can_execute("session1", "build")
    assert allowed is True


def test_invalid_session():
    se = StepEnforcer()
    se.add_step("deploy", ["build"])

    allowed, _ = se.can_execute("nonexistent", "deploy")
    assert allowed is False
